File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.java
class Solution { // Sliding Window
    public int maxProfit(int[] prices) {
        int left = 0; // But
        int right = 1; // Sell
        int maxProfit = 0;
        while (right < prices.length) {
                int currentProfit = prices[right] - prices[left];
                if (prices[left] < prices[right]) {
                    maxProfit = Math.max(currentProfit, maxProfit);
                }
                else {
                    left = right;
                }
                right += 1;
        }
        return maxProfit;
    }
}

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution:
    # Solution 1
    def maxProfit(self, prices: List[int]) -> int:
        maxProfit = 0
        lowest = prices[0]

        for price in prices:
            if price < lowest:
                lowest = price
            maxProfit = max(maxProfit, price - lowest)

        return maxProfit

    # Solution 2: Sliding Window
    def maxProfit2(self, prices: List[int]) -> int:
        l, r = 0, 1
        maxProfit = 0

        while r < len(prices):
            if prices[l] < prices[r]:
                profit = prices[r] - prices[l]
                maxProfit = max(maxProfit, profit)
            else:
                l = r
            r += 1
        return maxProfit